sphinx_needs/lsp/esbonio.py

```diff
@@ -0,0 +1,378 @@
+            new_path = "/".join(path.split("/")[current_path_length : current_path_length + 1])
+            if new_path not in sub_paths:
+                sub_paths.append(new_path)
+    sub_paths.sort()
+
+    items = []
+    for sub_path in sub_paths:
+        if sub_path.find(".rst") > -1:
+            kind = CompletionItemKind.File
+        else:
+            kind = 19  # Folder
+        items.append(CompletionItem(label=sub_path, kind=kind, detail="path to needs doc"))
+    return items
+
+
+def complete_need_link(ls, params: CompletionContext, lines: List[str], line: str, word: str):
+    # specify the need type, e.g.,
+    # ->req
+    if word.count(">") == 1:
+        return [CompletionItem(label=need_type, detail="need type") for need_type in ls.needs_store.types]
+
+    word_parts = word.split(">")
+
+    # specify doc in which need is specified, e.g.,
+    # ->req>fusion/index.rst
+    if word.count(">") == 2:
+        requested_type = word_parts[1]  # e.g., req, test, ...
+        if requested_type in ls.needs_store.types:
+            return doc_completion_items(ls, ls.needs_store.docs_per_type[requested_type], word_parts[2])
+
+    # specify the exact need, e.g.,
+    # ->req>fusion/index.rst>REQ_001
+    if word.count(">") == 3:
+        requested_type = word_parts[1]  # e.g., req, test, ...
+        requested_doc = word_parts[2]  # [0:-4]  # without `.rst` file extension
+        if requested_doc in ls.needs_store.needs_per_doc:
+            substitution = word[word.find("->") :]
+            start_char = line.find(substitution)
+            line_number = params.position.line
+            return [
+                CompletionItem(
+                    label=need["id"],
+                    insert_text=need["id"],
+                    documentation=need["description"],
+                    detail=need["title"],
+                    additional_text_edits=[
+                        TextEdit(
+                            range=Range(
+                                start=Position(line=line_number, character=start_char),
+                                end=Position(
+                                    line=line_number,
+                                    character=start_char + len(substitution),
+                                ),
+                            ),
+                            new_text="",
+                        )
+                    ],
+                )
+                for need in ls.needs_store.needs_per_doc[requested_doc]
+                if need["type"] == requested_type
+            ]
+
+
+def generate_hash(user_name, doc_uri, need_prefix, line_number):
+    salt = os.urandom(blake2b.SALT_SIZE)  # pylint: disable=no-member
+    return blake2b(
+        f"{user_name}{doc_uri}{need_prefix}{line_number}".encode(),
+        digest_size=4,
+        salt=salt,
+    ).hexdigest()
+
+
+def generate_need_id(ls, params, lines: List[str], word: str, need_type: str = None) -> str:
```

Comment:
mypy 'strict' checking is enabled, so i'm not sure why it's not flagging this one... 'strict' should imply 'no implicit optionals', so the signature here should be:

```suggestion
def generate_need_id(ls, params, lines: List[str], word: str, need_type: Optional[str] = None) -> str:
```

also `ls`, and `params` are missing type annotations